File: src/test/java/tech/simter/r2dbc/Sample.java
package tech.simter.r2dbc;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.PersistenceConstructor;
import org.springframework.data.domain.Persistable;
import org.springframework.lang.NonNull;

public class Sample implements Persistable<Integer> {
  @Id
  private final Integer id;
  private final String name;

  @PersistenceConstructor
  public Sample(@NonNull Integer id, @NonNull String name) {
    this.id = id;
    this.name = name;
  }

  public Sample withId(Integer id) {
    return id.equals(this.id) ? this : new Sample(id, this.name);
  }

  @Override
  public boolean isNew() {
    return true;
  }

  public Integer getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  @Override
  public boolean equals(Object obj) {
    if (!(obj instanceof Sample)) return false;
    Sample other = (Sample) obj;
    return this.id.equals(other.getId()) && this.name.equals(other.getName());
  }
}

File: src/test/kotlin/tech/simter/r2dbc/Sample.kt
package tech.simter.r2dbc

import org.springframework.data.annotation.Id
import org.springframework.data.domain.Persistable

data class Sample(
  @JvmField
  @Id val id: Int,
  val name: String
) : Persistable<Int> {
  override fun isNew(): Boolean {
    return true
  }

  override fun getId(): Int {
    return id
  }
}